summingbird-core/src/main/scala/com/twitter/summingbird/planner/DagOptimizer.scala

```diff
@@ -291,7 +291,7 @@ trait DagOptimizer[P <: Platform[P]] {
       // TODO: we need to case class here to not lose the irreducible which may be named
       case ValueFlatMappedProducer(in, fn) =>
         // we know that (K, V) <: T due to the case match, but scala can't see it
-        def cast[K, V](p: Prod[(K, V)]): Prod[T] = p.asInstanceOf[Prod[T]]
+        def cast[K, V](p: Prod[(K, V)]): Prod[T] = IdentityKeyedProducer[P, K, V](p).asInstanceOf[Prod[T]]
```

Comment:
I'm still not understanding why this is an error. Can you post a gist to a stack trace that hit the issue?